### ISSUE TITLE
Prediction with custom batch sampler

### ISSUE THREAD

**Issue by stecklin** [Lightning-AI/lightning#11807]
## 🐛 Bug

When using a custom batch sampler for prediction, the following code gets executed:

https://github.com/PyTorchLightning/pytorch-lightning/blob/ab1c2ff23fd27ab6e5647e390313b569e32b61c5/pytorch_lightning/trainer/data_loading.py#L221-L226

This assumes the custom batch sampler has the same interface as [BatchSampler](https://pytorch.org/docs/stable/data.html#torch.utils.data.BatchSampler). However, data loaders can take any [Sampler](https://pytorch.org/docs/stable/data.html#torch.utils.data.Sampler) as batch_sampler argument according to the docs: https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader

### To Reproduce

This is the custom sampler I was using:

```python
class MySampler(Sampler):
    def __init__(self, dataset):
        """
        Args:
            dataset (Dataset): PyTorch dataset.
        """
        self.dataset = dataset
        self.group_ids_to_sample_indices = self.get_group_ids_to_sample_indices()

    def get_group_ids_to_sample_indices(self):
        """ Group samples by id

        Returns:
            dict: group_id -> indices of samples with this id
        """
        group_ids_to_sample_indices = defaultdict(list)
        for i, sample in enumerate(self.dataset.samples):
            group_ids_to_sample_indices[sample["group_id"]].append(i)
        return group_ids_to_sample_indices

    def __iter__(self):
        group_ids = list(self.group_ids_to_sample_indices.keys())
        # Yield a batch of variable size with all samples for one group
        for group_id in group_ids:
            yield self.group_ids_to_sample_indices[group_id]

    def __len__(self):
        return len(self.group_ids_to_sample_indices)
```

### Expected behavior

Prediction runs without crashing. For training, validation and testing it already works.

### Environment

- PyTorch Lightning Version: 1.5.9
- PyTorch Version: 1.10.1

### Additional context

I did the following work around for now, but it feels very hacky:

```python
class MySampler(Sampler):
    def __init__(self, sampler=None, batch_size=None, drop_last=False, dataset=None):
        if sampler:
            self.dataset = sampler.data_source
        else:
            self.dataset = dataset
        self.group_ids_to_sample_indices = self.get_group_ids_to_sample_indices()
```

cc @tchaton @rohitgr7 @akihironitta @justusschock @awaelchli @ninginthecloud

**Comment by ninginthecloud** [comment]
Hi, @stecklin, thanks for flagging this issue! cc: @tchaton do we need to require batch_sampler to be `BatchSampler` type? we can get access of batch_size or drop_last via dataloader directly.

**Comment by tchaton** [comment]
Hey @stecklin @ninginthecloud, yes you are right. This code assumed the custom BatchSampler follows the BatchSampler API which can't be assumpted to be true. This code should be refactored and both index tracking, fault-tolerant won't activated for custom batch sampler

**Comment by denizetkar** [comment]
This problem absolutely breaks code that cannot afford to use the same interface as the `BatchSampler`. There isn't any workaround in such a case. So, may I ask, why is this issue not part of any milestone?

**Comment by carmocca** [comment]
I would do both of your suggestions, not touch it at all when we cannot do it and show a warning message in that case.

**Comment by awaelchli** [comment]
I agree with @carmocca. When users have custom objects, Lightning shouldn't touch them at all.

**Comment by otaj** [comment]
We already do touch them though - for example in case of DataLoaders in a hook in DDP setting

**Comment by TheShadow29** [comment]
Hello, I am having a similar issue. I have a custom batch sampler, but the required arguments are not passed because of this part of the code: https://github.com/Lightning-AI/lightning/blob/1.6.4/pytorch_lightning/utilities/data.py#L286. It restricts passing only batch_size, and drop_last.

I am using v1.6.4.

**Comment by TheShadow29** [comment]
Hi @otaj, Thanks for the interest. The batch sampler I am using is essentially a slightly modified version of `GroupedBatchSampler` here https://github.com/facebookresearch/detectron2/blob/main/detectron2/data/samplers/grouped_batch_sampler.py#L6. Basically, it allows all the items within a single batch to be from the same group. 

Since it requires the argument `group_ids`, the default version cannot be used.

For now, I have been able to get it working by setting `replace_sampler_ddp=False`, and making the dataloader with the distributed sampler stuff on my own. Happy to provide more details.

**Comment by otaj** [comment]
Hi @TheShadow29 , I opened a PR (#13640) yesterday that should fix this. I cannot make any guarantees on when it gets to the master, but we're getting there :)

**Comment by psridhar-asapp** [comment]
Hi I am also stuck on this. Pasting another example of a sampler that will break https://github.com/espnet/espnet/blob/master/espnet2/samplers/length_batch_sampler.py

This one doesn't even use the concept of batch_sizes. Instead it looks at something called batch_bins since it bins together sequences of similar length but varies the number of elements so that the total memory used is somewhat similar.

**Comment by psridhar-asapp** [comment]
I don't know if this will be helpful but I just forked your PR to test it and I'm running into the issue that I never enter this if branch at all so I get the same error with the default behaviour

https://github.com/Lightning-AI/lightning/blob/bugfix/custom_batch_sampler/src/pytorch_lightning/utilities/data.py#L328

**Comment by otaj** [comment]
Hi @psridhar-asapp, thank you for your report. To be fair, your `BatchSampler` will not work with my PR, because we are still doing two things, that are not satisfied by your sampler.

1. The PR is doing this sort of fix only for subclasses of `BatchSampler`, not a regular `Sampler`. While your `LengthBatchSampler` has a name `BatchSampler` in the name, it is in fact a subclass of `AbsSampler` which is in turn a subclass of `Sampler`. We might potentially consider doing similar type of fix for regular `Sampler`s, however, that would have to be much larger PR since Sampler are much more prevalent in PL codebase.
2. And while the PR is making PL to be more lenient wrt set of arguments custom `BatchSampler`s take, it still expects, that there is an argument called `Sampler` and without it, the PR will still raise an error (https://github.com/Lightning-AI/lightning/blob/9e106e12acf9ab30da22b9a611452164df0ecdec/src/pytorch_lightning/utilities/data.py#L345). Note that argument `batch_size` is now not required for custom `BatchSampler`s at all, but a `sampler` is still required
3. And lastly, while I cannot know if this is your case or not, all of this magic works only when the corresponding `BatchSampler` and `DataLoader` is instantiated in `*_dataloader` hook of your module.

**Comment by otaj** [comment]
But, my suggestion on how to make it all work.

1. Decouple loading of the data from your `Sampler`
2. Use very simple sampler, that produce a single example on every iteration. This could be `SequentialSampler` or `RandomSampler`, depending on your usecase. You don't really have to write your own solution here.
3. Modify your `LengthBatchSampler` to take this sampler, which produces the sequence of data as an argument called `sampler` and consume data from there instead of from your files.

And you're all set :tada: Hopefully, it won't be too difficult

**Comment by TheShadow29** [comment]
@otaj I wonder if you really need `BatchSampler` to have a `Sampler` since it already gives you the list of batch indices? So technically, you should be able to instantiate the sampler inside the batch_sampler, take multiple samplers as input, or have no sampler (dummy batch sampler) at all and just always output the same index.

For my current implementation: I have `replace_sampler_ddp=False`, and the batch_sampler gives me the list of indices for the entire batch (so batch_size_per_gpu * world_size). I then apply a `DistributedBatchSampler` (see below) on top of batch_sampler and pass it to the data loader.  

Let me know what you think?

```
# Adapted from here: 
# https://pytorchnlp.readthedocs.io/en/latest/_modules/torchnlp/samplers/distributed_batch_sampler.html

class DistributedBatchSampler(BatchSampler):
    def __init__(self, batch_sampler, **kwargs):
        self.batch_sampler = batch_sampler
        self.kwargs = kwargs

    def __iter__(self):
        for batch in self.batch_sampler:
            dist_sampler = DistributedSampler(batch, **self.kwargs)
            dist_sampler_lst = list(dist_sampler)
            batch_out = [batch[x] for x in dist_sampler_lst]
            yield batch_out

    def __len__(self):
        return len(self.batch_sampler)
```

**Comment by psridhar-asapp** [comment]
Hey @otaj to follow up, the sampler I pasted above does not load any training data, it merely loads precomputed sequence length files in order to better batch by length. I could move those files into a dataset and pass in a sequential sampler but this sampler would effectively be unused. During init I would still load the entire shape file and sort by length in order to build batches.

**Comment by otaj** [comment]
Hi, @TheShadow29, you're absolutely correct, that it is not strictly necessary to use `Sampler` inside your `BatchSampler`. However, the fact is, stock PyTorch uses `Sampler`s a lot and we cannot really know what is going on inside custom `BatchSampler`s. The PR allows for a custom `BatchSampler` given mild constrains.

And as for your code, I don't think it is necessary. PL allows you to use any `Sampler` as a distributed `Sampler` and if you wrap this `DistributedSampler` in a `BatchSampler`, you can batch them as you please :) But maybe there's something I'm missing

**Comment by otaj** [comment]
Hi @psridhar-asapp , I apologize, it seems I haven't read your sampler properly on Friday. Now I took a bit closer look and it seems you are still processing stream of indices which you are batching (https://github.com/espnet/espnet/blob/f2778f798b76e102602078213a21b63aa592be70/espnet2/samplers/length_batch_sampler.py#L98-L116) - this code could effectively be produced by sampler. However, I agree, that it would not be optimal then since you'd be doing the loading of the shapes twice.

However, this sampler will still work if you use `replace_sampler_ddp=False` in your Trainer. Supporting completely custom `BatchSampler`, which do not obey the protocol of `BatchSampler` would not be impossible for us, but is something, that the current PR does not do.

**Comment by otaj** [comment]
To shed a bit more light on why we don't support it at the moment - we might be doing changes to the underlying `Sampler` (such as making it distributed, unshuffled, etc.) and we want to reinstantiate `BatchSampler` with the new, changed `Sampler` in such a way, that the `BatchSampler` is aware of it. But it means, we got a `Sampler` instance, which we wouldn't know what to do with, if the `BatchSampler` doesn't support `Sampler`

**Comment by TheShadow29** [comment]
> And as for your code, I don't think it is necessary. PL allows you to use any `Sampler` as a distributed `Sampler` and if you wrap this `DistributedSampler` in a `BatchSampler`, you can batch them as you please :) But maybe there's something I'm missing

Hi @otaj, so the thing is if you directly use DistributedSampler over the Sampler inside BatchSampler, you run into using mis-matched groups across GPUs. Suppose there are two groups, group1 and group2. Samples from group2 take twice as long to compute than samples in group1. 

If I understand your solution correctly, making the sampler inside batchsampler could create the situation where GPU1 looks at group1 and GPU2 looks at group2. However, to sync the gradients for optimization, you would need to wait for group2 everytime leading to increased training time. 

The solution I suggested prevents that. It first creates the batch from a particular group (group1 or group2), and then distributes it across the gpus. So you ensure that each gpu gets same group samples. 

It is possible I am also missing something obvious. Let me know what you think.

**Comment by otaj** [comment]
@TheShadow29, you are correct that you want to be sure to prevent these ugly things such as waiting on syncing. However, `DistributedSampler` from PyTorch (which is what is used in PL) is no magic, where you wouldn't be able to control which samples are in which group. If you take a look here https://github.com/pytorch/pytorch/blob/master/torch/utils/data/distributed.py#L96, it is taking samples from the `Dataset` (in our case any `Sampler`) and distributes samples to every replica by slicing over original samples in a way of `indices = indices[self.rank:self.total_size:self.num_replicas]`. This is very much deterministic, if you turn off shuffling - which I believe you have to have turned off even for your scenario

**Comment by carmocca** [comment]
Fixed by https://github.com/Lightning-AI/lightning/pull/13640

**Comment by shenoynikhil** [comment]
Facing the same issue while using DynamicBatchSampler ([link](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/loader/dynamic_batch_sampler.html)) from `torch_geometric`. Any idea on how this can be prevented? @otaj

**Comment by vb123er951** [comment]
hi, does this problem get fixed? 
I have same problem that cannot use custom batch sampler during prediction.
the package version I am using are `pytorch-lightning 2.5.1` and  `torch 2.5.1`
